src/mcelog.c

```diff
@@ -0,0 +1,431 @@
+                if (fgets(buf, sizeof(buf), p_file) != NULL) {
+                  sscanf(buf, "\t%d total", &memory_record.corrected_err_total);
+                  DEBUG("%s: Got corrected error total %d", MCELOG_PLUGIN,
+                        memory_record.corrected_err_total);
+                }
+                if (fgets(buf, sizeof(buf), p_file) != NULL) {
+                  sscanf(buf, "\t%d in %s", &memory_record.corrected_err_timed,
+                         memory_record.corrected_err_timed_period);
+                  DEBUG("%s: Got timed corrected errors %d in %s",
+                        MCELOG_PLUGIN, memory_record.corrected_err_total,
+                        memory_record.corrected_err_timed_period);
+                }
+              }
+              if (!memcmp(buf, uncorrected_err, strlen(uncorrected_err))) {
+                if (fgets(buf, sizeof(buf), p_file) != NULL) {
+                  sscanf(buf, "\t%d total",
+                         &memory_record.uncorrected_err_total);
+                  DEBUG("%s: Got uncorrected error total %d", MCELOG_PLUGIN,
+                        memory_record.uncorrected_err_total);
+                }
+                if (fgets(buf, sizeof(buf), p_file) != NULL) {
+                  sscanf(buf, "\t%d in %s",
+                         &memory_record.uncorrected_err_timed,
+                         memory_record.uncorrected_err_timed_period);
+                  DEBUG("%s: Got timed uncorrected errors %d in %s",
+                        MCELOG_PLUGIN, memory_record.uncorrected_err_total,
+                        memory_record.uncorrected_err_timed_period);
+                }
+              }
+            }
+          }
+          fclose(p_file);
+        }
+      }
+    }
+  }
+  return NULL;
+}
+
+static int mcelog_init(void) {
+  if (!g_configured)
+    mcelog_config_init_default();
+
+  if (connect_to_client_socket() != 0) {
+    ERROR("%s: Cannot connect to client socket", MCELOG_PLUGIN);
+    return -1;
+  }
+
+  if (plugin_thread_create(&g_mcelog_config.tid, NULL, poll_worker, NULL) !=
+      0) {
+    ERROR("%s: Error creating poll thread.", MCELOG_PLUGIN);
+    return -1;
+  }
+  return 0;
+}
+
+static int get_memory_machine_checks(void) {
+  static const char dump[] = "dump all bios\n";
+
+  if (swrite(g_mcelog_config.sock_fd, dump, sizeof(dump)) < 0)
+    return (-1);
+
+  DEBUG("%s: SENT DUMP REQUEST", MCELOG_PLUGIN);
+  return 0;
+}
+
+static int mcelog_read(__attribute__((unused)) user_data_t *ud) {
+  DEBUG("%s: %s", MCELOG_PLUGIN, __FUNCTION__);
+
+  if (get_memory_machine_checks() != 0)
+    ERROR("%s: MACHINE CHECK INFO NOT AVAILABLE", MCELOG_PLUGIN);
+
+  return 0;
+}
+
+static int mcelog_shutdown(void) {
+  pthread_mutex_lock(&mcelog_thread_lock);
+  mcelog_thread_shutdown = 1;
+  pthread_mutex_unlock(&mcelog_thread_lock);
+  pthread_join(g_mcelog_config.tid, NULL);
+  pthread_mutex_destroy(&mcelog_thread_lock);
+  shutdown(g_mcelog_config.sock_fd, 0);
```

Comment:
Please check that `sock_fd` is in fact a valid file descriptor before calling `shutdown()`.